scalafix-core/shared/src/main/scala/scalafix/internal/rule/Disable.scala

```diff
@@ -21,25 +62,87 @@ final case class Disable(index: SemanticdbIndex, config: DisableConfig)
   override def description: String =
     "Linter that reports an error on a configurable set of symbols."
 
-  private lazy val disabledSymbol: SymbolMatcher =
-    SymbolMatcher.normalized(config.allSymbols: _*)
+  private lazy val disabledSymbolInSynthetics: SymbolMatcher =
+    SymbolMatcher.normalized(config.allSymbolsInSynthetics: _*)
+
+  private lazy val disabledBlock: SymbolMatcher =
+    SymbolMatcher.normalized(config.allUnlessBlocks: _*)
 
   override def init(config: Conf): Configured[Rule] =
     config
       .getOrElse("disable", "Disable")(DisableConfig.default)
       .map(Disable(index, _))
 
-  override def check(ctx: RuleCtx): Seq[LintMessage] = {
+  private def createLintMessage(
+      symbol: Symbol.Global,
+      signature: Signature,
+      custom: Option[CustomMessage[Symbol.Global]],
+      pos: Position,
+      details: String = ""): LintMessage = {
+    val message = config
+      .customMessage(symbol)
+      .flatMap(_.message)
+      .getOrElse(s"${signature.name} is disabled$details")
+
+    val id = custom
+      .flatMap(_.id)
+      .getOrElse(signature.name)
+
+    errorCategory
+      .copy(id = id)
+      .at(message, pos)
+  }
+
+  private def checkTree(ctx: RuleCtx): Seq[LintMessage] = {
+    def filterBlockedSymbolsInBlock(
+        blockedSymbols: List[Symbol.Global],
+        block: Tree): List[Symbol.Global] = {
+      val Some(symbolBlock: Symbol.Global) = ctx.index.symbol(block)
+      blockedSymbols.filter(sym =>
+        !config.symbolsInUnlessBlock(symbolBlock).contains(sym))
+    }
+
+    def treeIsBlocked(
+        tree: Tree,
+        blockedSymbols: List[Symbol.Global]): Boolean =
+      ctx.index.symbol(tree) match {
+        case Some(s: Symbol.Global) =>
+          blockedSymbols.exists(SymbolOps.isSameNormalized(_, s))
+        case _ => false
+      }
+
+    new SearcherWithContext(config.allSymbols)({
+      case (_: Import, _) => Right(List.empty)
+      case (
+          Term.Apply(Term.Select(disabledBlock(block), Term.Name("apply")), _),
+          blockedSymbols) =>
+        Right(filterBlockedSymbolsInBlock(blockedSymbols, block)) // <Block>.apply
+      case (Term.Apply(disabledBlock(block), _), blockedSymbols) =>
+        Right(filterBlockedSymbolsInBlock(blockedSymbols, block)) // <Block>(...)
+      case (_: Defn.Def, _) =>
+        Right(config.allSymbols) // reset blocked symbols in def
+      case (_: Term.Function, _) =>
+        Right(config.allSymbols) // reset blocked symbols in (...) => (...)
+      case (t: Name, blockedSymbols) if treeIsBlocked(t, blockedSymbols) =>
```

Comment:
As a general note, the scalafix API typically tries to make it possible to guard against stuff in the pattern instead of the guard.

Reply:
@olafurpg `blockedSymbols` isn't static, it changes during tree traversal.

Reply:
You can  construct a new SymbolMatcher whenever it changes https://github.com/scalacenter/scalafix/blob/master/scalafix-core/shared/src/main/scala/scalafix/util/SymbolMatcher.scala

Reply:
And what then? :) I receive a tuple `(Tree, List[Symbol.Global])` and want to test if `tree` is blocked by these symbols.

Reply:
```scala
val isBlocked = SymbolMatcher.normalized(badSymbols:_*)
tree match {
  case isBlocked(n: Name) =>
}
```
?

Reply:
:thinking: But why is it better than guard? I can place these code in some function :)

Reply:
Unless I'm seeing something wrong, it looks like `treeIsBlocked` is reimplementing the SymbolMatcher logic. You could replace it with SymbolMatcher and it would be clearer what is happening.

> It's even worse, I want to ignore cases when tree isn't blocked.

Why can't you ignore cases when the tree isn't blocked with SymbolMatcher?

Reply:
`val Some(...) = ...` is a sign that you were not able to prove to the compiler this deconstruction can't fail. By using `SymbolMatcher.unapply` you don't need to outsmart the typechecker.

Reply:
Yeah, I agree with `val Some(...) = ...`. 

>Why can't you ignore cases when the tree isn't blocked with SymbolMatcher?
 
Well, in `SearcherWithContext` I need to return new context or `LintError`, ignoring means no changing in the context. I could just return an old context... I didn't want to do this :)

Reply:
I'm not sure I follow why SymbolMatcher is not suitable here but I trust your judgment 😉 we can keep it unchanged

Reply:
I can change it to what you want, maybe it will be more understandable.

Reply:
Hehe, IDEA can't parse `case (ctx.index.Symbol(t: Name), ...`